sdk/cxx/src/message_dispatcher.cpp

```diff
@@ -0,0 +1,131 @@
+/*
+ Copyright 2017 Intel Corporation
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+------------------------------------------------------------------------------
+*/
+
+#include <exception>
+#include <iostream>
+#include <string>
+
+#include <zmq.hpp>
+
+#include "sawtooth/message_dispatcher.h"
+
+namespace sawtooth {
+
+MessageDispatcher::MessageDispatcher():
+        context_(1),
+        server_socket(this->context_, ZMQ_DEALER),
+        message_socket(this->context_, ZMQ_DEALER),
+        processessing_request_socket(this->context_, ZMQ_DEALER)
+        { }
+
+MessageDispatcher::~MessageDispatcher() {}
+
+void MessageDispatcher::Connect(const std::string& connection_string) {
+    std::cout << "MessageDispatcher::Connect" << std:: endl;
+    this->server_socket.connect(connection_string.c_str());
+    this->message_socket.bind("inproc://send_queue");
+    this->processessing_request_socket.bind("inproc://request_queue");
+
+    // start thread
+    std::unique_lock<std::mutex> lock(this->mutex);
+    this->dispatch_thread = std::thread(MessageDispatcher::SendReceiveThread,
+        this);
+
+    this->condition.wait(lock);
```

Comment:
Again, I think that it is recommended that you do:

```
while (!this->thread_is_ready) {
  this->condition.wait(lock);
}

```
Because waits can end for a reason other than the notiify()/notify_all() method being invoked on it.